### PR TITLE
Admin Page Prototype

### PR DESCRIPTION
**Pull request by amejia98**
-Created the Admin.jsp page, cannot be viewed by unregistered users or those who are not admins
   +The admin usernames are hardcoded for now
-Added unit tests for AdminServlet class
-Mapped the admin page on web.xml
-Updated teammates' names with correct spelling on the About page

### CODE REVIEW

File: src/main/java/codeu/controller/AdminServlet.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  /**
+   * Sets the UserStore used by this servlet. This function provides a common setup method for use
+   * by the test framework or the servlet's init() function.
+   */
+  void setUserStore(UserStore userStore) {
+    this.userStore = userStore;
+  }
+
+  /**
+   * This function fires when a user navigates to the admin page.
```
Comment (shershey) [inline src/main/java/codeu/controller/AdminServlet.java]: It's better to say what this does than what causes it to fire (though that can be useful to include near the end)  Something like:

Serves the Admin page for the logged in user.

TODO(amejia98 ): Will eventually do x, y, and z.

Users can navigate to the admin page either directly with (what's the URL?) or by clicking the admin link  from the (is there another way to get there).

File: src/main/java/codeu/controller/AdminServlet.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      request.setAttribute("non_admin", "You are not authorized to view this page");
+    } else {
+      request.setAttribute("admin", "Hi " + username + "! Welcome to the admin page!");
+    }
+
+    request.getRequestDispatcher("/WEB-INF/view/admin.jsp").forward(request, response);
+    return;
+  }
+
+  public boolean isAdmin(String username) {
```
Comment (shershey) [inline src/main/java/codeu/controller/AdminServlet.java]: Is this called elsewhere?  If not, you can make this private.

File: src/main/webapp/WEB-INF/view/admin.jsp [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE html>
+<html>
+<head>
+  <title>Admin Page</title>
+  <link rel="stylesheet" href="/css/main.css">
+</head>
+<body>
+
+	<nav>
```
Comment (shershey) [inline src/main/webapp/WEB-INF/view/admin.jsp]: Though there is no style guide for HTML, it's good to have the indentation match for opening and closing brackets and be consistent. Something like 2 spaces for indentation.

File: src/main/webapp/WEB-INF/view/admin.jsp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    <% } else{ %>
+      <a href="/login">Login</a>
+    <% } %>
+    <a href="/about.jsp">About</a>
+  </nav>
+
+  <div id="container">
+    <h1>Admin Page</h1>
+
+   <% if (request.getAttribute("unregistered_user") != null) { %>
```
Comment (shershey) [inline src/main/webapp/WEB-INF/view/admin.jsp]: I didn't catch this while reading the servlet code, but I think it's a little funny to have the servlet write the message.  I think, in a MVC (Model-view-controller) design, you want to put as much view in the JSP page as possible.  So I would instead do something like:

* Create a LoginState enum in the servlet
* Create a username attribute
* Use those two in the JSP file to decide what to say to the user.

File: src/main/webapp/WEB-INF/view/admin.jsp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        <h2><%= request.getAttribute("unregistered_user") %></h2>
+    <% } else if (request.getAttribute("non_admin") != null) { %>
+        <h2><%= request.getAttribute("non_admin") %></h2>
+    <% } else { %>
+        <h2><%= request.getAttribute("admin") %></h2>
+	  <% } %>
+  </div>
+
+</body>
+</html>
```
Comment (shershey) [inline src/main/webapp/WEB-INF/view/admin.jsp]: Github must have some linter.  It's complaining you don't have a newline at the end of your file

File: src/main/webapp/index.jsp [hunk 6]
```diff
@@ -46,6 +46,7 @@
             create or join a conversation.</li>
         <li>View the <a href="/about.jsp">about</a> page to learn more about the
             project.</li>
+        <li>View the <a href="/admin">admin</a> page for app-related statistics!</li>
```
Comment (shershey) [inline src/main/webapp/index.jsp]: In your final product, you might consider hiding this option from non admin users.  But probably better not to do that now.

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    mockResponse = Mockito.mock(HttpServletResponse.class);
+    mockRequestDispatcher = Mockito.mock(RequestDispatcher.class);
+    Mockito.when(mockRequest.getRequestDispatcher("/WEB-INF/view/admin.jsp"))
+        .thenReturn(mockRequestDispatcher);
+  }
+
+  // Tests accessing the admin page with an unregistered user
+  @Test
+  public void testDoGet_UnregisteredUserReturnsDeniedAccess() throws IOException, ServletException {
+  	Mockito.when(mockSession.getAttribute("user")).thenReturn(null);
```
Comment (shershey) [inline src/test/java/codeu/controller/AdminServletTest.java]: The indentation should be consistent and I believe 2 spaces: https://google.github.io/styleguide/javaguide.html#s4.2-block-indentation

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    	"You must login and be an admin to view this page.");
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+  }
+
+  // Tests accesing the admin page with a user who is not an admin. They should not be allowed access
+  @Test
+  public void testDoGet_NonAdminReturnsDeniedAccess() throws IOException, ServletException {
+    Mockito.when(mockSession.getAttribute("user")).thenReturn("test_username");
+  	
+  	adminServlet.doGet(mockRequest, mockResponse);
```
Comment (shershey) [inline src/test/java/codeu/controller/AdminServletTest.java]: Weird indentation

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+    Mockito.verify(mockRequest).setAttribute("non_admin", 
+    	"You are not authorized to view this page");
```
Comment (shershey) [inline src/test/java/codeu/controller/AdminServletTest.java]: Style guide says to use 4 spaces when continuing a line: https://google.github.io/styleguide/javaguide.html#s4.5.2-line-wrapping-indent

Also. This is super nitpicky but the style guide has some recommendations for where to line break:

https://google.github.io/styleguide/javaguide.html#s4.5.1-line-wrapping-where-to-break

So maybe:

Mockito.verify(mockRequest).setAttribute(
    "non_admin",  "You are not authorized to view this page");

Also, there's a 100 line limit.  If you're less than that you don't have to break.

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+  }
+
+  // Tests accessing the admin page with a user who is an admin 
+  @Test
+  public void testDoGet_AdminUserReturnsCanAccess() throws IOException, ServletException {
+    Mockito.when(mockSession.getAttribute("user")).thenReturn("amejia");
+  	
+  	adminServlet.doGet(mockRequest, mockResponse);
```
Comment (shershey) [inline src/test/java/codeu/controller/AdminServletTest.java]: Weird indentation

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    mockSession = Mockito.mock(HttpSession.class);
+    Mockito.when(mockRequest.getSession()).thenReturn(mockSession);
+    
+    mockResponse = Mockito.mock(HttpServletResponse.class);
+    mockRequestDispatcher = Mockito.mock(RequestDispatcher.class);
+    Mockito.when(mockRequest.getRequestDispatcher("/WEB-INF/view/admin.jsp"))
+        .thenReturn(mockRequestDispatcher);
+  }
+
+  // Tests accessing the admin page with an unregistered user
```
Comment (shershey) [inline src/test/java/codeu/controller/AdminServletTest.java]: These comments are probably unnecessary.  Google encourages only adding comments if they add additional information that the code does not imply.  You named your test methods well and so these comments are probably not necessary.

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+  public void testDoGet_AdminUserReturnsCanAccess() throws IOException, ServletException {
+    Mockito.when(mockSession.getAttribute("user")).thenReturn("amejia");
+  	
+  	adminServlet.doGet(mockRequest, mockResponse);
+
+    Mockito.verify(mockRequest).setAttribute("admin", "Hi amejia! Welcome to the admin page!");
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+  }
+
+}
```
Comment (shershey) [inline src/test/java/codeu/controller/AdminServletTest.java]: Add a newline.

File: src/test/java/codeu/controller/AdminServletTest.java (hunk 8 above)
Comment (amejia98) [reply src/test/java/codeu/controller/AdminServletTest.java]: Is there a reason why the indentation shows up different here than in my editor? I'm using Sublime and in the editor it looks like I am following the 2-space indentation rule. I attached a picture of what it looks like in the editor. How can I remedy this?
screen shot 2018-05-24 at 9 15 03 pm

Comment (amejia98) [reply src/test/java/codeu/controller/AdminServletTest.java]: Oops nevermind, found a solution online.

**Pull request closed without merging**